sympy/calculus/finite_diff.py

```diff
@@ -452,32 +450,51 @@ def differentiate_finite(expr, *symbols,
     >>> differentiate_finite(f(x)*g(x), x, points=[x-h, x+h])
     -f(-h + x)*g(-h + x)/(2*h) + f(h + x)*g(h + x)/(2*h)
 
-    Note that the above form preserves the product rule in discrete form.
-    If we want we can pass ``evaluate=True`` to get another form (which is
-    usually not what we want):
+    Note that the above form preserves the product rule in discrete form. If we
+    want we can pass ``evaluate=True`` to get another form (which is usually
+    not what we want):
 
     >>> differentiate_finite(f(x)*g(x), x, points=[x-h, x+h], evaluate=True).simplify()
     -((f(-h + x) - f(h + x))*g(x) + (g(-h + x) - g(h + x))*f(x))/(2*h)
 
-    ``differentiate_finite`` works on any expression:
+    ``differentiate_finite`` works on any expression, including expressions
+    with embedded derivatives that don't support ``evaluate=True``:
 
     >>> differentiate_finite(f(x) + sin(x), x, 2)
     -2*f(x) + f(x - 1) + f(x + 1) - 2*sin(x) + sin(x - 1) + sin(x + 1)
     >>> differentiate_finite(f(x) + sin(x), x, 2, evaluate=True)
     -2*f(x) + f(x - 1) + f(x + 1) - sin(x)
     >>> differentiate_finite(f(x, y), x, y)
     f(x - 1/2, y - 1/2) - f(x - 1/2, y + 1/2) - f(x + 1/2, y - 1/2) + f(x + 1/2, y + 1/2)
+    >>> differentiate_finite(f(x)*g(x).diff(x), x)
+    (-g(x) + g(x + 1))*f(x + 1/2) - (g(x) - g(x - 1))*f(x - 1/2)
+
+    To make finite difference with non-constant discretization step use
+    undefined functions:
+
+    >>> dx = Function('dx')
+    >>> differentiate_finite(f(x)*g(x).diff(x), points=dx(x))
+    -(-g(x - dx(x)/2 - dx(x - dx(x)/2)/2)/dx(x - dx(x)/2) + g(x - dx(x)/2 + dx(x - dx(x)/2)/2)/dx(x - dx(x)/2))*f(x - dx(x)/2)/dx(x) + (-g(x + dx(x)/2 - dx(x + dx(x)/2)/2)/dx(x + dx(x)/2) + g(x + dx(x)/2 + dx(x + dx(x)/2)/2)/dx(x + dx(x)/2))*f(x + dx(x)/2)/dx(x)
```

Comment:
You should be able to line wrap this.

Reply:
My thoughts exactly and if you look at the previous commits I tried to wrap it but the doctest was failing. The problem is that after 2a219e7bc4f7a5ecefc99f49128601d82b844e94 docstring `differentiate_finite.__doc__` was changed to raw string and line break `\` doesn't seem to work now. How does one wrap lines in the case of raw strings?

Another way would be to change the docstring to standard string again and use `\`. What do you think?

Reply:
Have you tried just wrapping it? The doctest runner should handle variable whitespace

Reply:
@oscarbenjamin I did, take a look at the previous commit c690901ea822cea. The line was wrapped as following
```python
    >>> dx = Function('dx')
    >>> differentiate_finite(f(x)*g(x).diff(x), points=dx(x))
    -(-g(x - dx(x)/2 - dx(x - dx(x)/2)/2)/dx(x - dx(x)/2) \
+ g(x - dx(x)/2 + dx(x - dx(x)/2)/2)/dx(x - dx(x)/2))*f(x - dx(x)/2)/dx(x) \
+ (-g(x + dx(x)/2 - dx(x + dx(x)/2)/2)/dx(x + dx(x)/2) \
+ g(x + dx(x)/2 + dx(x + dx(x)/2)/2)/dx(x + dx(x)/2))*f(x + dx(x)/2)/dx(x)
```
and the doctest failed. Is that what you meant?

Reply:
Have you tried without the slashes?

Reply:
@oscarbenjamin Yes, the latest commit e2e0f28d57b is without slashes but it is failing.

Is there possibility to move from raw string `r"""` to standard docstring `"""` than wrapping would work, would that be a possibility? Or are there any other ideas that might work?

Reply:
@asmeurer @oscarbenjamin The doc is wrapped now.

Reply:
I meant that you should be able to wrap it like this:
```diff
diff --git a/sympy/calculus/finite_diff.py b/sympy/calculus/finite_diff.py
index d6c24e72ea..4f62a42b04 100644
--- a/sympy/calculus/finite_diff.py
+++ b/sympy/calculus/finite_diff.py
@@ -474,10 +474,10 @@ def differentiate_finite(expr, *symbols,
 
     >>> dx = Function('dx')
     >>> differentiate_finite(f(x)*g(x).diff(x), points=dx(x))
-    -(-g(x - dx(x)/2 - dx(x - dx(x)/2)/2)/dx(x - dx(x)/2) + \
-g(x - dx(x)/2 + dx(x - dx(x)/2)/2)/dx(x - dx(x)/2))*f(x - dx(x)/2)/dx(x) + \
-(-g(x + dx(x)/2 - dx(x + dx(x)/2)/2)/dx(x + dx(x)/2) + \
-g(x + dx(x)/2 + dx(x + dx(x)/2)/2)/dx(x + dx(x)/2))*f(x + dx(x)/2)/dx(x)
+    -(-g(x - dx(x)/2 - dx(x - dx(x)/2)/2)/dx(x - dx(x)/2) +
+    g(x - dx(x)/2 + dx(x - dx(x)/2)/2)/dx(x - dx(x)/2))*f(x - dx(x)/2)/dx(x) +
+    (-g(x + dx(x)/2 - dx(x + dx(x)/2)/2)/dx(x + dx(x)/2) +
+    g(x + dx(x)/2 + dx(x + dx(x)/2)/2)/dx(x + dx(x)/2))*f(x + dx(x)/2)/dx(x)
 
     """
     # Key-word only arguments only available in Python 3
```
Whether it is a raw string or not shouldn't matter.